frontend/packages/console-app/src/components/cloud-shell/AutoResizeTerminal.tsx

```diff
@@ -0,0 +1,77 @@
+import * as React from 'react';
+import { Terminal as XTerminal } from 'xterm';
+import * as fit from 'xterm/lib/addons/fit/fit';
+
+XTerminal.applyAddon(fit);
+
+type TerminalProps = {
+  onData: (data: string) => void;
+  className?: string;
+};
+
+export class Terminal extends React.Component<TerminalProps> {
+  private terminalRef;
+
+  private terminal;
+
+  private resizeObserver;
+
+  constructor(props) {
+    super(props);
+    this.terminalRef = React.createRef();
+    const options = {
+      fontFamily: 'monospace',
+      fontSize: 16,
+      cursorBlink: false,
+      cols: 80,
+      rows: 25,
+      padding: 30,
+    };
+    this.terminal = new XTerminal(Object.assign({}, options));
+    this.terminal.on('data', this.props.onData);
+  }
+
+  focus = () => {
+    this.terminal && this.terminal.focus();
+  };
+
+  onResize() {
+    try {
+      this.terminal && this.terminal.fit();
+    } finally {
+      // do nothing
+    }
+  }
+
+  onDataReceived = (data) => {
+    this.terminal && this.terminal.write(data);
+  };
+
+  onConnectionClosed = (reason) => {
+    const { terminal } = this;
+    if (!terminal) {
+      return;
+    }
+    terminal.write(`\x1b[31m${reason || 'disconnected'}\x1b[m\r\n`);
+    terminal.cursorHidden = true;
+    terminal.setOption('disableStdin', true);
+    terminal.refresh(terminal.y, terminal.y);
+  };
+
+  componentDidMount() {
+    this.terminal.open(this.terminalRef.current);
+    this.focus();
+    this.resizeObserver = new ResizeObserver(() => {
+      this.terminal.fit();
+    });
+    this.resizeObserver.observe(this.terminalRef.current);
+  }
+
+  componentWillUnmount() {
+    this.terminal && this.terminal.destroy();
```

Comment:
You also need to disconnect the `resizeobserver` when component unmounts.
```suggestion
    this.terminal && this.terminal.destroy();
    this.resizeObserver && this.resizeObserver.disconnect();
```